pkg/resolver/multiresolver/multiresolver.go

```diff
@@ -0,0 +1,212 @@
+	}
+}
+
+// WithLogger will set the logger used by the MultiResolver.
+func WithLogger(logger logging.Logger) Option {
+	return func(mr *MultiResolver) {
+		mr.logger = logger
+	}
+}
+
+// WithForceDefault will force resolution using the default resolver chain.
+func WithForceDefault() Option {
+	return func(mr *MultiResolver) {
+		mr.ForceDefault = true
+	}
+}
+
+// PushResolver will push a new Resolver to the name resolution chain for the
+// given TLD. An empty TLD will push to the default resolver chain.
+func (mr *MultiResolver) PushResolver(tld string, r resolver.Interface) {
+	mr.resolvers[tld] = append(mr.resolvers[tld], r)
+}
+
+// PopResolver will pop the last reslover from the name resolution chain for the
+// given TLD. An empty TLD will pop from the default resolver chain.
+func (mr *MultiResolver) PopResolver(tld string) error {
+	l := len(mr.resolvers[tld])
+	if l == 0 {
+		return fmt.Errorf("tld %s: %w", tld, ErrResolverChainEmpty)
+	}
+	mr.resolvers[tld] = mr.resolvers[tld][:l-1]
+	return nil
+}
+
+// ChainCount retruns the number of resolvers in a resolver chain for the given
+// tld.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will
+// return the number of resolvers in the default resolver chain.
+func (mr *MultiResolver) ChainCount(tld string) int {
+	return len(mr.resolvers[tld])
+}
+
+// GetChain will return the resolution chain for a given TLD.
+// TLD names should be prepended with a dot (eg ".tld"). An empty TLD will
+// return all resolvers in the default resolver chain.
+func (mr *MultiResolver) GetChain(tld string) []resolver.Interface {
+	return mr.resolvers[tld]
+}
+
+// Resolve will attempt to resolve a name to an address.
+// The resolution chain is selected based on the TLD of the name. If the name
+// does not end in a TLD, the default resolution chain is selected.
+// The resolution will be performed iteratively on the resolution chain,
+// returning the result of the first Resolver that succeeds. If all resolvers
+// in the chain return an error, the function will return an ErrResolveFailed.
+func (mr *MultiResolver) Resolve(name string) (addr resolver.Address, err error) {
+	tld := ""
+	if !mr.ForceDefault {
+		tld = getTLD(name)
+	}
+	chain := mr.resolvers[tld]
+
+	// If no resolver chain is found, switch to the default chain.
+	if len(chain) == 0 {
+		chain = mr.resolvers[""]
+	}
+
+	errs := multierror.New()
+	for _, res := range chain {
+		addr, err = res.Resolve(name)
+		if err == nil {
+			return addr, nil
+		}
+		errs.Append(err)
+	}
+
+	return addr, errs.WrapErrorOrNil(
+		fmt.Errorf("resolve chain for tld %q: %w", tld, ErrResolverChainFailed))
+}
+
+// Close all will call Close on all resolvers in all resolver chains.
+func (mr *MultiResolver) Close() error {
+	errs := multierror.New()
+
+	for _, chain := range mr.resolvers {
+		for _, r := range chain {
+			if err := r.Close(); err != nil {
+				errs.Append(err)
+			}
+		}
+	}
+
+	return errs.WrapErrorOrNil(ErrCloseFailed)
```

Comment:
I am not sure if ErrCloseFailed is needed, as the caller already knows that this is a Close method call that it returned error. It is a redundant information.

Reply:
Does the caller need to filter using `errors.Is`, specifically for `Close`-type errors? Clients may return different errors, this should wrap it at the top level.

If, however, this is not idiomatic (or applicable to the project) - I'm happy to remove it.

Reply:
Since this error is only returned by the Close method, `errors.Is` is not needed, `err != nil` gives the same information that there has been an error. Clients can return different errors, but I see not point to have a special error like ErrCloseFailed if that error is returned by only one method. I do not see that it provides any new information.

Reply:
Removed.